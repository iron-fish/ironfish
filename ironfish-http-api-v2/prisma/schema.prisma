generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model accounts {
  id             Int      @id @default(autoincrement())
  public_address String   @db.VarChar
  created_at     DateTime @default(now()) @db.Timestamp(6)
  updated_at     DateTime @default(now()) @db.Timestamp(6)
  events         events[]

  @@index([public_address], name: "index_accounts_on_public_address")
}

model events {
  id         Int        @id @default(autoincrement())
  created_at DateTime   @default(now()) @db.Timestamp(6)
  updated_at DateTime   @default(now()) @db.Timestamp(6)
  account_id Int
  type       event_type
  accounts   accounts   @relation(fields: [account_id], references: [id])

  @@index([account_id], name: "index_events_on_account_id")
}

enum event_type {
  BLOCK_MINED
  BUG_CAUGHT
  COMMUNITY_CONTRIBUTION
  NODE_HOSTED
  PULL_REQUEST_MERGED
  SOCIAL_MEDIA_PROMOTION
}
