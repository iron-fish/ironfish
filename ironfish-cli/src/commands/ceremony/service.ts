/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { createRootLogger, setLogPrefixFromConfig } from '@ironfish/sdk'
import { Flags } from '@oclif/core'
import { IronfishCommand } from '../../command'
import { RemoteFlags } from '../../flags'
import { CeremonyServer } from '../../trusted-setup/server'
import { S3Utils } from '../../utils'

const CONTRIBUTE_TIMEOUT_MS = 5 * 60 * 1000
const UPLOAD_TIMEOUT_MS = 5 * 60 * 1000
const PRESIGNED_EXPIRATION_SEC = 5 * 60
const START_DATE = 1680904800000 // Friday, April 07 2023 15:00:00 GMT-0700 (Pacific Daylight Time)

export default class CeremonyService extends IronfishCommand {
  static hidden = true

  static description = `
     Start the coordination server for the Iron Fish trusted setup ceremony
   `

  static flags = {
    ...RemoteFlags,
    bucket: Flags.string({
      char: 'b',
      parse: (input: string) => Promise.resolve(input.trim()),
      required: false,
      description: 'S3/R2 bucket to download and upload params to',
      default: 'ironfish-contributions',
    }),
    downloadPrefix: Flags.string({
      char: 'b',
      parse: (input: string) => Promise.resolve(input.trim()),
      required: false,
      description: 'Prefix for contribution download URLs',
      default: 'https://contributions.ironfish.network',
    }),
    contributionTimeoutMs: Flags.integer({
      required: false,
      description: 'Allowable milliseconds for a contributor to run the contribution script',
      default: CONTRIBUTE_TIMEOUT_MS,
    }),
    uploadTimeoutMs: Flags.integer({
      required: false,
      description: 'Allowable milliseconds for a contributor to upload their new parameters',
      default: UPLOAD_TIMEOUT_MS,
    }),
    presignedExpirationSec: Flags.integer({
      required: false,
      description:
        'How many seconds the S3/R2 pre-signed upload URL is valid for a contributor',
      default: PRESIGNED_EXPIRATION_SEC,
    }),
    startDate: Flags.integer({
      required: false,
      description: 'When should the server start accepting contributions',
      default: START_DATE,
    }),
    token: Flags.string({
      required: true,
    }),
    skipIPCheck: Flags.boolean({
      required: false,
      description: 'Pass this flag if you want to skip checking for duplicate IPs',
      default: false,
    }),
  }

  async start(): Promise<void> {
    const { flags } = await this.parse(CeremonyService)

    const DEFAULT_HOST = '0.0.0.0'
    const DEFAULT_PORT = 9040

    const r2Credentials = await S3Utils.getR2Credentials('us-east-1')

    if (r2Credentials === undefined) {
      this.logger.log('Failed getting R2 credentials from AWS')
      this.exit(0)
      return
    }

    const r2Client = S3Utils.getR2S3Client(r2Credentials)

    setLogPrefixFromConfig(`[%tag%]`)

    const server = new CeremonyServer({
      logger: createRootLogger(),
      port: DEFAULT_PORT,
      host: DEFAULT_HOST,
      s3Bucket: flags.bucket,
      downloadPrefix: flags.downloadPrefix,
      s3Client: r2Client,
      tempDir: this.sdk.config.tempDir,
      contributionTimeoutMs: flags.contributionTimeoutMs,
      uploadTimeoutMs: flags.uploadTimeoutMs,
      presignedExpirationSec: flags.presignedExpirationSec,
      startDate: flags.startDate,
      token: flags.token,
      enableIPBanning: !flags.skipIPCheck,
    })

    await server.start()

    await server.waitForStop()
  }
}
