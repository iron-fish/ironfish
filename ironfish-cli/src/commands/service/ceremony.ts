/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Flags } from '@oclif/core'
import { IronfishCommand } from '../../command'
import { RemoteFlags } from '../../flags'
import { CeremonyServer } from '../../trusted-setup/server'

export default class Ceremony extends IronfishCommand {
  static hidden = true

  static description = `
     Start the coordination server for the Iron Fish trusted setup ceremony
   `

  static flags = {
    ...RemoteFlags,
    bucket: Flags.string({
      char: 'b',
      parse: (input: string) => Promise.resolve(input.trim()),
      required: false,
      description: 'S3 bucket to download and upload params to',
      default: 'ironfish-snapshots',
    }),
  }

  async start(): Promise<void> {
    const { flags } = await this.parse(Ceremony)

    const DEFAULT_HOST = '0.0.0.0'
    const DEFAULT_PORT = 9040

    const server = new CeremonyServer({
      logger: this.logger,
      port: DEFAULT_PORT,
      host: DEFAULT_HOST,
    })

    server.start()

    await server.waitForStop()
  }
}
