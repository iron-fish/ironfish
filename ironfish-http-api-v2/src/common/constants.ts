/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
export const MS_PER_DAY = 1000 * 60 * 60 * 24
