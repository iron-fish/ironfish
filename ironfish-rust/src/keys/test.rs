/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

use crate::keys::{ephemeral::EphemeralKeyPair, PUBLIC_ADDRESS_SIZE};

use super::{shared_secret, PublicAddress, SaplingKey};
use group::Curve;
use jubjub::ExtendedPoint;

#[test]
fn test_key_generation_and_construction() {
    let key = SaplingKey::generate_key();
    let key2 = SaplingKey::new(key.spending_key).unwrap();
    assert!(key.spending_key != [0; 32]);
    assert!(key2.spending_key == key.spending_key);
    assert!(key2.incoming_viewing_key.view_key == key.incoming_viewing_key.view_key);
}

#[test]
fn test_diffie_hellman_shared_key() {
    let key1 = SaplingKey::generate_key();

    let address1 = key1.public_address();

    let key_pair = EphemeralKeyPair::new();
    let secret_key = key_pair.secret();
    let public_key = key_pair.public();

    let shared_secret1 = shared_secret(secret_key, &address1.transmission_key, public_key);
    let shared_secret2 = shared_secret(&key1.incoming_viewing_key.view_key, public_key, public_key);
    assert_eq!(shared_secret1, shared_secret2);
}

#[test]
fn test_diffie_hellman_shared_key_with_other_key() {
    let key = SaplingKey::generate_key();
    let third_party_key = SaplingKey::generate_key();

    let address = key.public_address();
    let third_party_address = third_party_key.public_address();

    let key_pair = EphemeralKeyPair::new();
    let secret_key = key_pair.secret();
    let public_key = key_pair.public();

    let shared_secret1 = shared_secret(secret_key, &address.transmission_key, public_key);
    let shared_secret2 = shared_secret(&key.incoming_viewing_key.view_key, public_key, public_key);
    assert_eq!(shared_secret1, shared_secret2);

    let shared_secret_third_party1 = shared_secret(
        secret_key,
        &third_party_address.transmission_key,
        public_key,
    );
    assert_ne!(shared_secret1, shared_secret_third_party1);
    assert_ne!(shared_secret2, shared_secret_third_party1);

    let shared_secret_third_party2 = shared_secret(
        &third_party_key.incoming_viewing_key.view_key,
        public_key,
        public_key,
    );
    assert_ne!(shared_secret1, shared_secret_third_party2);
    assert_ne!(shared_secret2, shared_secret_third_party2);
}

#[test]
fn test_serialization() {
    let key = SaplingKey::generate_key();
    let mut serialized_key = [0; PUBLIC_ADDRESS_SIZE];
    key.write(&mut serialized_key[..])
        .expect("Should be able to serialize key");
    assert_ne!(serialized_key, [0; PUBLIC_ADDRESS_SIZE]);

    let read_back_key = SaplingKey::read(&mut serialized_key.as_ref())
        .expect("Should be able to load key from valid bytes");
    assert_eq!(
        read_back_key.incoming_view_key().view_key,
        key.incoming_view_key().view_key
    );

    let public_address = key.public_address();
    let mut serialized_address = [0; PUBLIC_ADDRESS_SIZE];
    public_address
        .write(&mut serialized_address[..])
        .expect("should be able to serialize address");

    let read_back_address: PublicAddress = PublicAddress::new(&serialized_address)
        .expect("Should be able to construct address from valid bytes");

    assert_eq!(
        ExtendedPoint::from(read_back_address.transmission_key).to_affine(),
        ExtendedPoint::from(public_address.transmission_key).to_affine()
    )
}

#[test]
fn test_hex_conversion() {
    let key = SaplingKey::generate_key();

    let hex = key.hex_spending_key();
    assert_eq!(hex.len(), 64);
    let second_key = SaplingKey::from_hex(&hex).unwrap();
    assert_eq!(second_key.spending_key, key.spending_key);

    let address = key.public_address();
    let hex = address.hex_public_address();
    assert_eq!(hex.len(), 2 * PUBLIC_ADDRESS_SIZE);
    let second_address = PublicAddress::from_hex(&hex).unwrap();
    assert_eq!(second_address, address);

    assert!(PublicAddress::from_hex("invalid").is_err());
}

#[test]
fn test_from_and_to_words() {
    let key_bytes = [
        213, 107, 36, 28, 169, 101, 179, 153, 116, 133, 204, 240, 100, 33, 116, 12, 29, 97, 22, 57,
        34, 173, 28, 2, 238, 105, 251, 224, 146, 83, 218, 247,
    ];
    let words_for_bytes = "step float already fan forest smile spirit ridge vacant canal fringe blouse stock mention tonight fiber bright blast omit water ankle clarify hint turn".to_owned();

    // Convert to words
    let key = SaplingKey::new(key_bytes).expect("Key should be created");
    let mnemonic = key
        .to_words(bip39::Language::English)
        .expect("Should return words");
    let words = mnemonic.into_phrase();
    assert_eq!(words_for_bytes, words);

    // Convert from words
    let key =
        SaplingKey::from_words(words, bip39::Language::English).expect("key should be created");
    assert_eq!(key.spending_key, key_bytes);
}
