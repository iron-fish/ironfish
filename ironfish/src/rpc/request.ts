/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Event } from '../event'

export class RpcRequest<TRequest = unknown, TResponse = unknown> {
  data: TRequest
  route: string
  ended = false
  closed = false
  onEnd: (status: number, data?: TResponse) => void
  onStream: (data?: TResponse) => void
  onClose = new Event<[]>()

  constructor(
    data: TRequest,
    route: string,
    onEnd: (status: number, data?: unknown) => void,
    onStream: (data?: unknown) => void,
  ) {
    this.data = data
    this.route = route
    this.onEnd = onEnd
    this.onStream = onStream
  }

  end(data?: TResponse, status?: number): void {
    if (this.ended) {
      throw new Error(`Request has already ended`)
    }
    this.ended = true
    if (this.closed) {
      return
    }
    this.onClose.clear()
    this.onEnd(status || 200, data)
  }

  stream(data: TResponse): void {
    if (this.closed) {
      return
    }
    if (this.ended) {
      throw new Error(`Request has already ended`)
    }
    this.onStream(data)
  }

  close(): void {
    this.closed = true
    this.onClose.emit()
  }
}
