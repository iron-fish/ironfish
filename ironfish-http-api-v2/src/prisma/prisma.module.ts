/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common'
import { PrismaService } from './prisma.service'

@Module({
  exports: [PrismaService],
  providers: [PrismaService],
})
export class PrismaModule {}
