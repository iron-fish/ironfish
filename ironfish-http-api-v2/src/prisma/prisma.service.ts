/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common'
import { PrismaClient } from '@prisma/client'

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit, OnModuleDestroy {
  async onModuleInit(): Promise<void> {
    await this.$connect()
  }

  async onModuleDestroy(): Promise<void> {
    await this.$disconnect()
  }
}
