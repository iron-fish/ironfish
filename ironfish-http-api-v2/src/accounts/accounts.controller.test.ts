/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common'
import request from 'supertest'
import { v4 as uuid } from 'uuid'
import { PrismaService } from '../prisma/prisma.service'
import { bootstrapTestApp } from '../test/test-app'

describe('AccountsController', () => {
  let app: INestApplication
  let prisma: PrismaService

  beforeAll(async () => {
    app = await bootstrapTestApp()
    prisma = app.get(PrismaService)
    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  describe('GET /accounts/:id', () => {
    describe('with a valid id', () => {
      it('returns the account', async () => {
        const account = await prisma.account.create({
          data: {
            public_address: uuid(),
          },
        })
        const { body } = await request(app.getHttpServer())
          .get(`/accounts/${account.id}`)
          .expect(HttpStatus.OK)

        expect(body).toMatchObject({
          id: account.id,
          public_address: account.public_address,
        })
      })
    })

    describe('with a missing id', () => {
      it('returns a 404', async () => {
        await request(app.getHttpServer()).get('/accounts/123').expect(HttpStatus.NOT_FOUND)
      })
    })
  })
})
