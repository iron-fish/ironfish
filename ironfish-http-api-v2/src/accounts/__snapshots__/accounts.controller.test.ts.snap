// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsController GET /accounts/:id/metrics with a TOTAL request and no time range returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Must provide time range for \\"TOTAL\\" requests",
  "statusCode": 422,
}
`;

exports[`AccountsController GET /accounts/:id/metrics with a missing granularity returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "granularity must be a valid enum value",
  ],
  "statusCode": 422,
}
`;

exports[`AccountsController GET /accounts/:id/metrics with a time range for a LIFETIME request returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Cannot provide time range for \\"LIFETIME\\" requests",
  "statusCode": 422,
}
`;

exports[`AccountsController GET /accounts/:id/metrics with a time range longer than the supported range returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "Time range too long",
  "statusCode": 422,
}
`;

exports[`AccountsController GET /accounts/:id/metrics with end but no start returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "start must be a Date instance",
  ],
  "statusCode": 422,
}
`;

exports[`AccountsController GET /accounts/:id/metrics with start after end returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "\\"start\\" must be stricly less than \\"end\\"",
  "statusCode": 422,
}
`;

exports[`AccountsController GET /accounts/:id/metrics with start but no end returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "end must be a Date instance",
  ],
  "statusCode": 422,
}
`;
